guiEditor/src/components/sceneExplorer/treeItemSelectableComponent.tsx

```diff
@@ -0,0 +1,158 @@
+    filter: Nullable<string>
+}
+
+export class TreeItemSelectableComponent extends React.Component<ITreeItemSelectableComponentProps, { isExpanded: boolean, isSelected: boolean }> {
+    private _wasSelected = false;
+
+    constructor(props: ITreeItemSelectableComponentProps) {
+        super(props);
+
+        this.state = { isSelected: this.props.entity === this.props.selectedEntity, isExpanded: this.props.mustExpand ||  Tools.LookForItem(this.props.entity, this.props.selectedEntity) };
+    }
+
+    switchExpandedState(): void {
+        this.setState({ isExpanded: !this.state.isExpanded });
+    }
+
+    shouldComponentUpdate(nextProps: ITreeItemSelectableComponentProps, nextState: { isExpanded: boolean, isSelected: boolean }) {
+        if (!nextState.isExpanded && this.state.isExpanded) {
+            return true;
+        }
+
+        if (nextProps.selectedEntity) {
+            if (nextProps.entity === nextProps.selectedEntity) {
+                nextState.isSelected = true;
+                return true;
+            } else {
+                nextState.isSelected = false;
+            }
+
+            if (Tools.LookForItem(nextProps.entity, nextProps.selectedEntity)) {
+                nextState.isExpanded = true;
+                return true;
+            }
+        }
+
+        return true;
+    }
+
+    scrollIntoView() {
+        const element = ReactDOM.findDOMNode(this) as Element;
+
+        if (element) {
+            element.scrollIntoView(false);
+        }
+    }
+
+    componentDidMount() {
+        if (this.state.isSelected) {
+            this.scrollIntoView();
+        }
+    }
+
+    componentDidUpdate() {
+        if (this.state.isSelected && !this._wasSelected) {
+            this.scrollIntoView();
+        }
+        this._wasSelected = false;
+    }
+
+    onSelect() {
+        if (!this.props.globalState.onSelectionChangedObservable) {
+            return;
+        }
+        this._wasSelected = true;
+        const entity = this.props.entity;
+        this.props.globalState.onSelectionChangedObservable.notifyObservers(entity);
+    }
+
+    renderChildren() {
+        const entity = this.props.entity;
+        if (!entity.getChildren && !entity.children || !this.state.isExpanded) {
+            return null;
+        }
+
+        const children = Tools.SortAndFilter(entity, entity.getChildren ? entity.getChildren() : entity.children);
+        return (
+            children.map((item, i) => {
+
+                return (
+                    <TreeItemSelectableComponent globalState={this.props.globalState} mustExpand={this.props.mustExpand} extensibilityGroups={this.props.extensibilityGroups} selectedEntity={this.props.selectedEntity} 
+                    key={i} offset={this.props.offset + 2} entity={item} filter={this.props.filter} />
+                );
+            })
+        )
+    }
+
+    render() {
+        const marginStyle = {
+            paddingLeft: (10 * (this.props.offset + 0.5)) + "px"

```

Comment:
is that the expected padding calculation? (just wondering about the `0.5`, which will certainly cause non-integer values to be set)

Reply:
Yes for now. This will get changed with the scene explorer visual changes next.